File: client/presenters/storyGraph.ts
import type { Hero, Place, Quest, Subject } from '@/types/domain';

export type StoryGraphNodeKind = 'hero' | 'place' | 'subject' | 'quest' | 'location' | 'target';

const NODE_KIND = new Set<StoryGraphNodeKind>([
  'hero',
  'place',
  'subject',
  'quest',
  'location',
  'target',
]);

export type StoryGraphNode = {
  id: string;
  kind: StoryGraphNodeKind;
  label: string;
  detail: string;
};

export type StoryGraphEdge = {
  id: string;
  source: string;
  target: string;
  label: string;
};

export type StoryGraphModel = {
  nodes: StoryGraphNode[];
  edges: StoryGraphEdge[];
  summary: string;
};

export const EMPTY_STORY_GRAPH: StoryGraphModel = {
  nodes: [],
  edges: [],
  summary: '스토리 데이터 없음',
};

type StoryGraphInput = {
  hero: Hero | null;
  subject: Subject | null;
  quest: Quest | null;
  place: Place | null;
};

const KIND_LABEL: Record<StoryGraphNodeKind, string> = {
  hero: '주인공',
  place: '현재 위치',
  subject: '대상',
  quest: '퀘스트',
  location: '배경',
  target: '등장인물',
};

function slug(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[^a-z0-9가-힣_-]/g, '');
}

function nodeId(kind: StoryGraphNodeKind, label: string): string {
  return `${kind}:${slug(label) || 'unknown'}`;
}

function characterId(label: string): string {
  return `character:${slug(label) || 'unknown'}`;
}

function shortList(items: string[], fallback = '정보 없음'): string {
  if (items.length === 0) return fallback;
  return items.slice(0, 3).join(', ');
}

function statLine(level: number, raceJob: string): string {
  return `Lv.${level} · ${raceJob}`;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function stringField(source: Record<string, unknown>, keys: string[]): string | null {
  for (const key of keys) {
    const value = source[key];
    if (typeof value === 'string' && value.trim()) return value;
    if (typeof value === 'number') return String(value);
  }
  return null;
}

function kindField(source: Record<string, unknown>): StoryGraphNodeKind {
  const raw = stringField(source, ['kind', 'type', 'entity_type', 'category']);
  if (raw === 'character') return 'target';
  if (raw === 'npc') return 'target';
  if (raw === 'player') return 'hero';
  if (raw && NODE_KIND.has(raw as StoryGraphNodeKind)) return raw as StoryGraphNodeKind;
  return 'target';
}

export function buildStoryGraph({ hero, subject, quest, place }: StoryGraphInput): StoryGraphModel {
  const nodes = new Map<string, StoryGraphNode>();
  const edges = new Map<string, StoryGraphEdge>();

  const addNode = (node: StoryGraphNode) => {
    nodes.set(node.id, node);
    return node.id;
  };

  const addEdge = (source: string | null, target: string | null, label: string) => {
    if (!source || !target || source === target) return;
    const id = `${source}->${target}:${label}`;
    edges.set(id, { id, source, target, label });
  };

  const heroId = hero
    ? addNode({
        id: nodeId('hero', hero.name),
        kind: 'hero',
        label: hero.name,
        detail: statLine(hero.level, hero.raceJob),
      })
    : null;

  const placeId = place
    ? addNode({
        id: nodeId('place', place.name),
        kind: 'place',
        label: place.name,
        detail: shortList([place.dayPhase, ...place.weather, place.risk.label].filter(Boolean)),
      })
    : null;

  const subjectId = subject
    ? addNode({
        id: characterId(subject.name),
        kind: 'subject',
        label: subject.name,
        detail: `${statLine(subject.level, subject.raceJob)} · ${subject.role}`,
      })
    : null;

  const questId = quest
    ? addNode({
        id: nodeId('quest', quest.title),
        kind: 'quest',
        label: quest.title,
        detail: `${quest.difficulty} · 보상 ${quest.rewards.gold}G/${quest.rewards.exp}EXP`,
      })
    : null;

  addEdge(heroId, placeId, '현재 위치');
  addEdge(heroId, subjectId, '주시');
  addEdge(subjectId, placeId, '같은 장소');
  addEdge(heroId, questId, '진행 중');

  if (questId && subjectId && subject && quest?.giver === subject.name) {
    addEdge(subjectId, questId, '의뢰');
  }

  for (const surrounding of place?.surroundings ?? []) {
    const locationId = addNode({
      id: nodeId('location', surrounding.name),
      kind: 'location',
      label: surrounding.name,
      detail: surrounding.risk.label,
    });
    addEdge(placeId, locationId, '이동');
  }

  for (const target of place?.targets ?? []) {
    const targetId =
      target.name === subject?.name && subjectId
        ? subjectId
        : addNode({
            id: characterId(target.name),
            kind: 'target',
            label: target.name,
            detail: statLine(target.level, target.raceJob),
          });
    addEdge(targetId, placeId, '등장');
  }

  const countByKind = Array.from(nodes.values()).reduce(
    (acc, node) => ({ ...acc, [node.kind]: acc[node.kind] + 1 }),
    {
      hero: 0,
      place: 0,
      subject: 0,
      quest: 0,
      location: 0,
      target: 0,
    } satisfies Record<StoryGraphNodeKind, number>,
  );

  const heroSummary = hero ? (hero.name === '주인공' ? '주인공' : `주인공 ${hero.name}`) : null;

  const summary = [
    heroSummary,
    place ? `현재 위치 ${place.name}` : null,
    quest ? `퀘스트 ${quest.title}` : null,
    `${KIND_LABEL.target} ${countByKind.target + countByKind.subject}`,
    `${KIND_LABEL.location} ${countByKind.location + countByKind.place}`,
  ]
    .filter(Boolean)
    .join(' · ');

  return {
    nodes: Array.from(nodes.values()),
    edges: Array.from(edges.values()),
    summary,
  };
}

export function mergeStoryGraphs(base: StoryGraphModel, next: StoryGraphModel): StoryGraphModel {
  if (base.nodes.length === 0 && base.edges.length === 0) return next;
  if (next.nodes.length === 0 && next.edges.length === 0) return base;

  const nextPlaceIds = new Set(next.nodes.filter((node) => node.kind === 'place').map((node) => node.id));
  const nextSubjectIds = new Set(next.nodes.filter((node) => node.kind === 'subject').map((node) => node.id));
  const nodes = new Map(base.nodes.map((node) => [node.id, node]));
  for (const [id, node] of nodes) {
    if (node.kind === 'place' && !nextPlaceIds.has(id)) {
      nodes.set(id, { ...node, kind: 'location' });
    }
    if (node.kind === 'subject' && !nextSubjectIds.has(id)) {
      nodes.set(id, { ...node, kind: 'target' });
    }
  }
  for (const node of next.nodes) {
    nodes.set(node.id, node);
  }

  const edges = new Map(base.edges.map((edge) => [edge.id, edge]));
  for (const edge of next.edges) {
    edges.set(edge.id, edge);
  }

  return {
    nodes: Array.from(nodes.values()),
    edges: Array.from(edges.values()),
    summary: next.summary,
  };
}

export function storyGraphFingerprint(graph: StoryGraphModel): string {
  const nodes = graph.nodes
    .map((node) => `${node.id}:${node.kind}:${node.label}:${node.detail}`)
    .sort()
    .join('|');
  const edges = graph.edges
    .map((edge) => `${edge.id}:${edge.source}:${edge.target}:${edge.label}`)
    .sort()
    .join('|');
  return `${graph.summary}::${nodes}::${edges}`;
}

export function normalizeStoryGraphPayload(payload: unknown): StoryGraphModel | null {
  const root = isObject(payload) && isObject(payload.graph) ? payload.graph : payload;
  if (!isObject(root) || !Array.isArray(root.nodes) || !Array.isArray(root.edges)) return null;

  const nodes = root.nodes.flatMap((raw): StoryGraphNode[] => {
    if (!isObject(raw)) return [];
    const id = stringField(raw, ['id', 'node_id', 'key']);
    const label = stringField(raw, ['label', 'name', 'title']);
    if (!id || !label) return [];
    const kind = kindField(raw);
    return [{
      id,
      kind,
      label,
      detail: stringField(raw, ['detail', 'description', 'summary', 'subtitle']) ?? KIND_LABEL[kind],
    }];
  });

  const knownNodeIds = new Set(nodes.map((node) => node.id));
  const edges = root.edges.flatMap((raw): StoryGraphEdge[] => {
    if (!isObject(raw)) return [];
    const source = stringField(raw, ['source', 'source_id', 'from', 'from_id']);
    const target = stringField(raw, ['target', 'target_id', 'to', 'to_id']);
    if (!source || !target || !knownNodeIds.has(source) || !knownNodeIds.has(target)) return [];
    const label = stringField(raw, ['label', 'type', 'relation']) ?? '관계';
    return [{
      id: stringField(raw, ['id']) ?? `${source}->${target}:${label}`,
      source,
      target,
      label,
    }];
  });

  return {
    nodes,
    edges,
    summary: stringField(root, ['summary', 'title']) ?? `서버 그래프 · ${nodes.length}N/${edges.length}E`,
  };
}

File: server/src/mapping/story_graph.py
"""Story-graph projection for the client's map view; characters are limited to player + current scene + companions + active subject so it can't become a spoiler dump."""

from typing import Literal

from ..domain.clock import day_phase
from ..domain.entities import Character, Location, Quest
from ..domain.state import GameState
from ..ontology.graph import GameGraph
from ..ontology.queries import (
    companions_of,
    connections_of,
    giver_of,
    inhabitants_of,
    location_of,
    trigger_targets_of,
)
from .labels import (
    RISK_PAYLOAD,
    gender_label,
    giver_with_location_label,
    race_job_label,
)


NodeStatus = Literal[
    "current",
    "engaged",
    "reachable_move",
    "reachable_meet",
    "unreachable_move",
    "unreachable_meet",
]
EdgeKind = Literal[
    "current_pin",
    "observe",
    "progress",
    "move",
    "meet",
    "quest_giver",
    "quest_target",
]


def _edge_id(source: str, target: str, label: str) -> str:
    return f"{source}->{target}:{label}"


def _reachable_location_ids(
    state: GameState, graph: GameGraph, start_location_id: str | None
) -> set[str]:
    if start_location_id is None or start_location_id not in state.locations:
        return set()
    seen = {start_location_id}
    queue = [start_location_id]
    while queue:
        location_id = queue.pop(0)
        for edge in connections_of(graph, location_id):
            target_id = edge.to_id
            if target_id not in state.locations or target_id in seen:
                continue
            seen.add(target_id)
            queue.append(target_id)
    return seen


def _visible_character_ids(
    state: GameState, graph: GameGraph, player: Character | None
) -> set[str]:
    if player is None:
        return set()

    visible = {player.id}
    if state.active_subject_id in state.characters:
        visible.add(state.active_subject_id)
    visible.update(
        cid for cid in companions_of(graph, player.id) if cid in state.characters
    )

    if player.location_id is not None:
        for cid in inhabitants_of(graph, player.location_id):
            if cid in state.characters:
                visible.add(cid)

    return visible


def _character_identity(state: GameState, graph: GameGraph, char: Character) -> dict:
    """Identity fields shared by every character-kind node (hero / subject /
    target). `role` and kind-specific extras (`known`, `trust`, `status`,
    `reachable`) are left to each builder — target uses appearance for its
    role label rather than `char.role`, hero has no `trust` against itself."""
    return {
        "id": char.id,
        "label": char.name,
        "level": char.level,
        "raceJob": race_job_label(state, graph, char),
        "gender": gender_label(char),
    }


def _hero_node(state: GameState, graph: GameGraph, player: Character) -> dict:
    return {
        **_character_identity(state, graph, player),
        "kind": "hero",
        "status": None,
        "reachable": True,
        "alive": player.alive,
        "role": player.role,
        "known": [player.appearance] if player.appearance else [],
    }


def _subject_node(
    state: GameState, graph: GameGraph, subject: Character, player: Character
) -> dict:
    known = [subject.appearance] if subject.appearance else []
    known += [m.content for m in player.memories if m.target_id == subject.id]
    return {
        **_character_identity(state, graph, subject),
        "kind": "subject",
        "status": "engaged",
        "reachable": True,
        "alive": subject.alive,
        "role": subject.role,
        "trust": subject.relations.get(player.id, 0),
        "known": known,
    }


def _target_node(
    state: GameState,
    graph: GameGraph,
    target: Character,
    player: Character,
    same_location: bool,
) -> dict:
    return {
        **_character_identity(state, graph, target),
        "kind": "target",
        "status": "reachable_meet" if same_location else "unreachable_meet",
        "reachable": same_location,
        "alive": target.alive,
        "role": target.appearance or target.description,
        "trust": target.relations.get(player.id, 0),
    }


def _place_node(loc: Location, state: GameState) -> dict:
    return {
        "id": loc.id,
        "kind": "place",
        "label": loc.name,
        "status": "current",
        "reachable": True,
        "description": loc.description,
        "risk": RISK_PAYLOAD[loc.sleep_risk],
        "dayPhase": day_phase(state.turn_count),
        "weather": list(loc.weather),
    }


def _location_node(loc: Location, move_difficulty: str | None, adjacent: bool) -> dict:
    return {
        "id": loc.id,
        "kind": "location",
        "label": loc.name,
        "status": "reachable_move" if adjacent else "unreachable_move",
        "reachable": adjacent,
        "description": loc.description,
        "risk": RISK_PAYLOAD[loc.sleep_risk],
        "moveDifficulty": move_difficulty,
    }


def _quest_node(quest: Quest, state: GameState, graph: GameGraph) -> dict:
    return {
        "id": quest.id,
        "kind": "quest",
        "label": quest.title,
        "status": None,
        "reachable": True,
        "questDifficulty": quest.difficulty,
        "rewards": {"gold": quest.rewards.gold, "exp": quest.rewards.exp},
        "giver": giver_with_location_label(state, graph, quest.id),
        "goals": [t.name for t in quest.triggers],
        "summary": quest.summary or "",
    }


def to_story_graph(state: GameState, graph: GameGraph | None = None) -> dict:
    """Project the server-side state into the client's story graph contract.

    Each node carries the kind-specific fields the panel renders
    (level/raceJob/risk/...); the panel reads them directly without
    re-deriving from a display string.
    """
    if graph is None:
        graph = state.graph()

    nodes: dict[str, dict] = {}
    edges: dict[str, dict] = {}
    player = state.characters.get(state.player_id)
    player_location_id = player.location_id if player else None
    visible_location_ids = _reachable_location_ids(state, graph, player_location_id)
    visible_character_ids = _visible_character_ids(state, graph, player)

    # Adjacent (one-hop) move difficulties keyed by neighbor location id —
    # used to populate `moveDifficulty` on adjacent location nodes only.
    # Multi-hop reachable locations get null; the panel only acts on
    # one-hop neighbors anyway.
    adjacent_move_difficulty: dict[str, str | None] = {}
    if player_location_id in state.locations:
        for edge in connections_of(graph, player_location_id):
            attrs = edge.attrs or {}
            adjacent_move_difficulty[edge.to_id] = attrs.get("difficulty")

    def add_edge(
        source: str | None, target: str | None, label: str, kind: EdgeKind
    ) -> None:
        if not source or not target or source == target:
            return
        if source not in nodes or target not in nodes:
            return
        edge_id = _edge_id(source, target, label)
        edges[edge_id] = {
            "id": edge_id,
            "source": source,
            "target": target,
            "label": label,
            "kind": kind,
        }

    if player is not None and player.id in visible_character_ids:
        nodes[player.id] = _hero_node(state, graph, player)

    adjacent_location_ids = set(adjacent_move_difficulty.keys())
    for location_id, location in state.locations.items():
        if location_id not in visible_location_ids:
            continue
        if location_id == player_location_id:
            nodes[location_id] = _place_node(location, state)
        else:
            nodes[location_id] = _location_node(
                location,
                adjacent_move_difficulty.get(location_id),
                adjacent=location_id in adjacent_location_ids,
            )

    for character_id in visible_character_ids:
        if character_id == state.player_id:
            continue
        character = state.characters.get(character_id)
        if character is None or player is None:
            continue
        if character_id == state.active_subject_id:
            nodes[character_id] = _subject_node(state, graph, character, player)
        else:
            same_location = location_of(graph, character_id) == player_location_id
            nodes[character_id] = _target_node(
                state, graph, character, player, same_location=same_location
            )

    quest_ids = (
        {state.active_quest_id} if state.active_quest_id in state.quests else set()
    )
    for quest_id in quest_ids:
        nodes[quest_id] = _quest_node(state.quests[quest_id], state, graph)

    add_edge(state.player_id, player_location_id, "현재 위치", "current_pin")
    add_edge(state.player_id, state.active_subject_id, "주시", "observe")
    add_edge(state.player_id, state.active_quest_id, "진행 중", "progress")

    for location_id in visible_location_ids:
        for edge in connections_of(graph, location_id):
            add_edge(location_id, edge.to_id, "이동", "move")

    for character_id in visible_character_ids:
        if character_id == state.player_id:
            continue
        loc_id = location_of(graph, character_id)
        if loc_id == player_location_id:
            add_edge(character_id, loc_id, "등장", "meet")

    for quest_id in quest_ids:
        add_edge(giver_of(graph, quest_id), quest_id, "의뢰", "quest_giver")
        for target_id in trigger_targets_of(graph, quest_id):
            add_edge(target_id, quest_id, "목표", "quest_target")

    count_by_kind: dict[str, int] = {
        "hero": 0,
        "place": 0,
        "location": 0,
        "subject": 0,
        "target": 0,
        "quest": 0,
    }
    for node in nodes.values():
        count_by_kind[node["kind"]] = count_by_kind.get(node["kind"], 0) + 1

    active_quest = (
        state.quests.get(state.active_quest_id) if state.active_quest_id else None
    )
    summary = " · ".join(
        part
        for part in [
            "주인공" if player is not None else None,
            f"현재 위치 {state.locations[player_location_id].name}"
            if player_location_id in state.locations
            else None,
            f"퀘스트 {active_quest.title}" if active_quest is not None else None,
            f"등장인물 {count_by_kind['target'] + count_by_kind['subject']}",
            f"장소 {count_by_kind['location'] + count_by_kind['place']}",
        ]
        if part
    )

    return {
        "nodes": list(nodes.values()),
        "edges": list(edges.values()),
        "summary": summary or "스토리 데이터 없음",
    }
